src/node_sets.cpp

```diff
@@ -0,0 +1,317 @@
+#include <algorithm>  // std::transform
+#include <cmath>
+#include <fmt/format.h>
+#include <sstream>
+#include <type_traits>
+
+#include <json.hpp>
+
+#include <bbp/sonata/node_sets.h>
+
+namespace bbp {
+namespace sonata {
+
+namespace {
+using json = nlohmann::json;
+
+const size_t MAX_COMPOUND_RECURSION = 10;
+
+template <typename T>
+std::string toString(const std::string& key, const std::vector<T>& values) {
+    return fmt::format(R"("{}": [{}])", key, fmt::join(values, ", "));
+}
+
+template <>
+std::string toString(const std::string& key, const std::vector<std::string>& values) {
+    // strings need to be wrapped in quotes
+    return fmt::format(R"("{}": ["{}"])", key, fmt::join(values, "\", \""));
+}
+
+template <typename T>
+class NodeSetBasicRule: public NodeSetRule
+{
+  public:
+    NodeSetBasicRule(std::string attribute, std::vector<T>&& values)
+        : attribute_(std::move(attribute))
+        , values_(values) {}
+
+    Selection materialize(const NodeSets& /* unused */, const NodePopulation& np) const final {
+        Selection ret{{}};
+        for (const auto& v : values_) {
+            ret = ret | np.matchAttributeValues(attribute_, v);
+        }
+        return ret;
+    }
+
+    std::string toJSON() const final {
+        return toString(attribute_, values_);
+    }
+
+  private:
+    std::string attribute_;
+    std::vector<T> values_;
+};
+
+class NodeSetBasicPopulation: public NodeSetRule
+{
+  public:
+    explicit NodeSetBasicPopulation(std::string name)
```

Comment:
should be names actually. 
To me there is no reason why you cannot have a list of populations with this : 
``` 
{
"pop1" : {"population" : "population1"},
"pop12" : {"population" : ["population1", "population2"]},
}
```
Allowing this would also the possibility to : 
``` 
{
"pop1" : {"population" : ["population1"]},
}
```
without raising

Reply:
sure